### PR TITLE
Make socksify and rack gem dependencies optional

### LINKED ISSUE #205: Licence issue with Socksify gem
The socksify dependancy makes httpi not usable for commercial projects since it's using GPL licence. Or am I wrong? :)

### LINKED ISSUE #224: Can we remove socksify as a dependency?
https://github.com/savonrb/httpi/blob/497a304e5d66f3670890f9611745395ecbd1cd3b/httpi.gemspec#L19

It would be awesome if we can make this an optional dependency because we are concerned about thread safety issues it might introduce - and also it looks abandoned.

### PR DESCRIPTION
**Pull request by ioquatix**
Fixes #224. Fixes #205.

### CODE REVIEW

File: .github/workflows/development.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            experimental: true
+          - os: ubuntu
+            ruby: head
+            experimental: true
+    
+    steps:
+    - uses: actions/checkout@v2
+    
+    - name: Install dependencies
+      run: apt-get install curl
```
Comment (olleolleolle) [inline .github/workflows/development.yml:40]: Don't it need sudo? Well, you are making this, great, I'll stop making the GitHub Actions PR I was fiddling with.

File: lib/httpi/adapter/rack.rb [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (olleolleolle) [inline lib/httpi/adapter/rack.rb:1]:
```suggestion
```

Typo?

**Review by olleolleolle — COMMENTED** [review]
Thanks!

File: .github/workflows/development.yml (hunk 1 above)
Comment (olleolleolle) [reply .github/workflows/development.yml:40]:
```suggestion
      run: sudo apt-get install libcurl-dev
```

Do we get headers with that other package?

**Comment by olleolleolle** [comment]
I am frantically pushing APPROVE AND RUN

File: README.md [hunk 3]
```diff
@@ -49,6 +39,20 @@ HTTPI.adapter = :httpclient
 HTTPI.request(:custom, request)
 ```
 
+### Rack Mock Adapter
+
+To use the Rack mock adaptor, please add the `rack` gem to your gemfile.
```
Comment (olleolleolle) [inline README.md:44]:
```suggestion
To use the Rack mock adapter, please add the `rack` gem to your gemfile.
```

File: spec/httpi/httpi_spec.rb [hunk 4]
```diff
@@ -294,7 +292,7 @@
       end
 
       HTTPI::Adapter::ADAPTERS.each do |adapter, opts|
-        unless (adapter == :em_http && RUBY_VERSION =~ /1\.8/) || (adapter == :curb && RUBY_PLATFORM =~ /java/)
+        unless (adapter == :em_http || adapter == :curb) && RUBY_PLATFORM =~ /java/
```
Comment (olleolleolle) [inline spec/httpi/httpi_spec.rb:295]: As mentioned, a platform-to-adapters-supported-on-that-platform could be interesting.

Comment (ioquatix) [reply spec/httpi/httpi_spec.rb:295]: Yeah this was a quick hack to get it passing since it's already 1AM lol.

Comment (olleolleolle) [reply spec/httpi/httpi_spec.rb:295]: All in due time, that's a note for future PRs.

**Comment by olleolleolle** [comment]
So, I think this is at the place where we need to look at things we can extract from it, and offer to the default branch.

A way forward with this big a change is a new major version, which allows to shed old version support etc.

**Comment by olleolleolle** [comment]
In addition to the earlier things fixed, the CI issue is _also_ fixed by this.

**Comment by pcai** [comment]
Thank you for your work on this.

File: spec/httpi/httpi_spec.rb (hunk 4 above)
Comment (Gobombers) [reply spec/httpi/httpi_spec.rb]: Yes

Comment (Gobombers) [reply spec/httpi/httpi_spec.rb]: Yes

**Comment by ioquatix** [comment]
Thank you so much.

**Pull request merged**